Scheduler - new Dev UI - first iteration

mkouba opened the pull request.

mkouba committed:
Scheduler - new Dev UI - first iteration

Co-authored-by: Phillip Krüger <phillip.kruger@gmail.com>

gsmet commented on extensions/scheduler/deployment/src/main/java/io/quarkus/scheduler/deployment/devui/SchedulerDevUIProcessor.java:
```diff
@@ -0,0 +1,33 @@
+package io.quarkus.scheduler.deployment.devui;
```
Probably a stupid question but we don't have to remove anything from the old UI?

Or you plan to clean that up later?

mkouba replied:
I have no idea what the plan is but it looks like we keep the old UI for now. CC @phillip-kruger

gsmet replied:
OK I thought we would remove the old UI as we add to the new as, otherwise, it will be extra work in dev mode.

But that can be done in the end if we think it's better.

phillip-kruger replied:
Yes we will remove the old Dev UI once the new one is usable

mkouba committed:
Scheduler - a more defensive design of SchedulerContext synthetic bean

- SchedulerContext#getScheduledMethods() should return immutable
metadata; it's an internal bean but can still be injected and modified

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+import { JsonRpc } from 'jsonrpc';
+import '@vaadin/grid';
+import '@vaadin/text-field';
+import '@vaadin/vertical-layout';
```
```suggestion
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+    jsonRpc = new JsonRpc("Scheduler");
+
+    static styles = css`
+        .schedules-table {
```
```suggestion
       :host {
            display: flex;
            flex-direction: column;
            gap: 10px;
        }
    
        .schedules-table {
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+
+        .annotation {
+          color: var(--lumo-contrast-50pct);
+        }
```
```suggestion
        }
        vaadin-button {
            cursor:pointer;
        }
    
        .topBar {
            display: flex;
            justify-content: space-between;
        }
        .searchField {
            width: 30%;
            padding-left: 20px;
        }
        .scheduler {
            padding-right: 20px;
        }
        
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+        if (this._scheduledMethods){
+            let schedulerButton;
+            if (this._schedulerRunning) {
+                schedulerButton = html`<vaadin-button @click=${() => this._pauseScheduler()}>
```
```suggestion
                schedulerButton = html`<vaadin-button class="scheduler" theme="tertiary" @click=${() => this._pauseScheduler()}>
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+                <vaadin-icon icon="font-awesome-solid:circle-pause"></vaadin-icon>
+                Pause scheduler</vaadin-button>`;
+            } else {
+                schedulerButton = html`<vaadin-button @click=${() => this._resumeScheduler()}>
```
```suggestion
                schedulerButton = html`<vaadin-button class="scheduler" theme="tertiary" @click=${() => this._resumeScheduler()}>
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+            }
+            
+            const searchBox = html`
+            <vaadin-text-field
+                placeholder="Search"
+                style="width: 30%;"
```
```suggestion
            <vaadin-text-field class="searchField"
                placeholder="Search"
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+            `
+            
+            return html`
+                <vaadin-vertical-layout theme="spacing">
+                    ${schedulerButton}
+                    ${searchBox}
```
```suggestion
                <div class="topBar">
                    ${searchBox}
                    ${schedulerButton}
                </div>
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+                        resizable>
+                    </vaadin-grid-column>
+                </vaadin-grid>
+                </vaadin-vertical-layout>
```
```suggestion
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+        let trigger;
+        if (schedule.identity) {
+            if (schedule.running) {
+                trigger = html`<vaadin-button @click=${() => this._pauseJob(schedule.identity)}>
```
```suggestion
                trigger = html`<vaadin-button theme="small" @click=${() => this._pauseJob(schedule.identity)}>
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+                <vaadin-icon icon="font-awesome-solid:circle-pause"></vaadin-icon>
+                Pause</vaadin-button>`;
+            } else {
+                trigger = html`<vaadin-button @click=${() => this._resumeJob(schedule.identity)}>
```
```suggestion
                trigger = html`<vaadin-button theme="small" @click=${() => this._resumeJob(schedule.identity)}>
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+
+    _methodRenderer(scheduledMethod) {
+      return html`
+        <vaadin-button @click=${() => this._executeMethod(scheduledMethod.methodDescription)}>
```
```suggestion
        <vaadin-button theme="small" @click=${() => this._executeMethod(scheduledMethod.methodDescription)}>
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+        this.jsonRpc.pauseJob({"identity": identity}).then(jsonResponse => {
+            if (jsonResponse.result.success) {
+                this._updateIdentityStatus(identity, false);
+                notifier.showSuccessMessage(jsonResponse.result.message, "bottom-stretch");
```
```suggestion
                notifier.showSuccessMessage(jsonResponse.result.message);
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+                this._updateIdentityStatus(identity, false);
+                notifier.showSuccessMessage(jsonResponse.result.message, "bottom-stretch");
+            } else {
+                notifier.showErrorMessage(jsonResponse.result.message, "bottom-stretch");
```
```suggestion
                notifier.showErrorMessage(jsonResponse.result.message);
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+        this.jsonRpc.resumeJob({"identity": identity}).then(jsonResponse => {
+            if (jsonResponse.result.success) {
+                this._updateIdentityStatus(identity, true);
+                notifier.showSuccessMessage(jsonResponse.result.message, "bottom-stretch");
```
```suggestion
                notifier.showSuccessMessage(jsonResponse.result.message);
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+        this.jsonRpc.pauseScheduler().then(jsonResponse => {
+            if (jsonResponse.result.success) {
+                this._schedulerRunning = false;
+                notifier.showSuccessMessage(jsonResponse.result.message, "bottom-stretch");
```
```suggestion
                notifier.showSuccessMessage(jsonResponse.result.message);
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+                this._schedulerRunning = false;
+                notifier.showSuccessMessage(jsonResponse.result.message, "bottom-stretch");
+            } else {
+                notifier.showErrorMessage(jsonResponse.result.message, "bottom-stretch");
```
```suggestion
                notifier.showErrorMessage(jsonResponse.result.message);
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+        this.jsonRpc.resumeScheduler().then(jsonResponse => {
+            if (jsonResponse.result.success) {
+                this._schedulerRunning = true;
+                notifier.showSuccessMessage(jsonResponse.result.message, "bottom-stretch");
```
```suggestion
                notifier.showSuccessMessage(jsonResponse.result.message);
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+                this._schedulerRunning = true;
+                notifier.showSuccessMessage(jsonResponse.result.message, "bottom-stretch");
+            } else {
+                notifier.showErrorMessage(jsonResponse.result.message, "bottom-stretch");
```
```suggestion
                notifier.showErrorMessage(jsonResponse.result.message);
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+    _executeMethod(methodDescription) {
+        this.jsonRpc.executeJob({"methodDescription": methodDescription}).then(jsonResponse => {
+            if (jsonResponse.result.success) {
+                notifier.showSuccessMessage(jsonResponse.result.message, "bottom-stretch");
```
```suggestion
                notifier.showSuccessMessage(jsonResponse.result.message);
```

phillip-kruger commented on extensions/scheduler/deployment/src/main/resources/dev-ui/scheduler/qwc-scheduler-scheduled-methods.js:
```diff
@@ -0,0 +1,251 @@
+            if (jsonResponse.result.success) {
+                notifier.showSuccessMessage(jsonResponse.result.message, "bottom-stretch");
+            } else {
+                notifier.showErrorMessage(jsonResponse.result.message, "bottom-stretch");
```
```suggestion
                notifier.showErrorMessage(jsonResponse.result.message);
```

phillip-kruger reviewed (changes requested):
Hi @mkouba This looks great ! I made a few cosmetic changes (suggestions) - see the screenshots in the comment below. I also did not understand what is being streamed to the client. I can see the dom change every couple of seconds, but nothing really change ? Also (as a separate effort) I think this extension is a good use case for a log tab ? A schedule log that logs all timeouts ? w.d.t.y ?

Before: 
![scheduler_current](https://user-images.githubusercontent.com/6836179/222303269-0fbce0d3-37b8-4500-8f10-67bd45508bdc.png)

After my suggested changes (just cosmetics)
![scheduler_proposed](https://user-images.githubusercontent.com/6836179/222303352-ea7fbaf2-2380-472a-9b76-3afba180781c.png)

mkouba commented:
@phillip-kruger I've applied your suggestions but I don't think it's practical to position the "Pause/Resume scheduler" button to the right. It's hard to spot.

phillip-kruger reviewed (approved):
LGTM ! Thanks !

phillip-kruger commented:
> I have no idea what "log tab" means. Is it covered in a video in your Quarkus Dev UI playlist?

Yes ! https://youtu.be/lN1Vjza9jm8

> Also what exactly do you mean with timeout? A scheduled job being executed?

Exactly. So it's easy to see that the scheduler execute.
